File: src/main/java/org/openbase/planetsudo/game/strategy/NoPlan.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.openbase.planetsudo.game.strategy;

/*-
 * #%L
 * PlanetSudo GameEngine
 * %%
 * Copyright (C) 2009 - 2018 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import org.openbase.planetsudo.level.levelobjects.Agent;
import org.openbase.planetsudo.level.levelobjects.AgentInterface;

/**
 *
 * @author <a href="mailto:divine@openbase.org">Divine Threepwood</a>
 */
public class NoPlan extends AbstractStrategy {

	public NoPlan(AgentInterface agent) {
		super(agent);
	}

	/**
	 * Hier wird angegeben wie viele Agenten dem Team zur Verfügung stehen sollen.
	 * @return Anzahl der Agenten
	 */
	@Override
	protected int loadAgentCount() {
		return 4;
	}

	@Override
	protected void loadRules() {
		//-------------------------------------------->
		createRule(new Rule(0, "Just Go") {
			@ Override
			protected boolean constraint() {
				return true;
			}
			@ Override
			protected void action() {
				agent.go();
				//agent.
			}
		});
		//-------------------------------------------->
		createRule(new Rule(999, "Wall") {
			@ Override
			protected boolean constraint() {
				return agent.isCollisionDetected();
			}
			@ Override
			protected void action() {
				agent.turnRight(95);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(500, "Flee") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryAgent();
			}
			@ Override
			protected void action() {
				if(agent.isCarringResource())
					agent.releaseResource();
				agent.turnRight(90);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(12, "Harvesting") {
			@ Override
			protected boolean constraint() {
				return agent.isCarringResource();
			}
			@ Override
			protected void action() {
				if(agent.isAtMothership()){
					agent.deliverResourceToMothership();
				}else{
					agent.goToMothership();
				}
			}
		});
		//-------------------------------------------->
		createRule(new Rule(10, "Collecting") {
			@ Override
			protected boolean constraint() {
				return agent.isTouchingResource();
			}
			@ Override
			protected void action() {
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(9, "MoveToRess") {
			@ Override
			protected boolean constraint() {
				return agent.seeResource();
			}
			@ Override
			protected void action() {
				agent.goToResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(13, "Help") {
			@ Override
			protected boolean constraint() {
				return agent.seeLostTeamAgent();
			}
			@ Override
			protected void action() {
				agent.spendTeamAgentFuel(3);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(800, "Fuel") {
			@ Override
			protected boolean constraint() {
				return agent.isAtMothership() && (agent.getFuelInPercent()<=50);
			}
			@ Override
			protected void action() {
				agent.orderFuel(100);
			}
		});
	}
}

File: src/main/kotlin/org/openbase/planetsudo/game/strategy/NoPlan.kt
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.openbase.planetsudo.game.strategy

import org.openbase.planetsudo.level.levelobjects.AgentInterface

/**
 *
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */
class NoPlan(agent: AgentInterface) : AbstractStrategy(agent) {
    /**
     * Hier wird angegeben wie viele Agenten dem Team zur Verfügung stehen sollen.
     * @return Anzahl der Agenten
     */
    override fun loadAgentCount(): Int {
        return 4
    }

    override fun loadRules() {
        // -------------------------------------------->
        createRule(object : Rule(0, "Just Go") {
            override fun constraint(): Boolean {
                return true
            }

            override fun action() {
                agent.go()
                // agent.
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(999, "Wall") {
            override fun constraint(): Boolean {
                return agent.isCollisionDetected
            }

            override fun action() {
                agent.turnRight(95)
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(500, "Flee") {
            override fun constraint(): Boolean {
                return agent.seeAdversaryAgent()
            }

            override fun action() {
                if (agent.isCarryingResource) agent.releaseResource()
                agent.turnRight(90)
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(12, "Harvesting") {
            override fun constraint(): Boolean {
                return agent.isCarryingResource
            }

            override fun action() {
                if (agent.isAtMothership) {
                    agent.deliverResourceToMothership()
                } else {
                    agent.goToMothership()
                }
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(10, "Collecting") {
            override fun constraint(): Boolean {
                return agent.isTouchingResource
            }

            override fun action() {
                agent.pickupResource()
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(9, "MoveToRess") {
            override fun constraint(): Boolean {
                return agent.seeResource()
            }

            override fun action() {
                agent.goToResource()
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(13, "Help") {
            override fun constraint(): Boolean {
                return agent.seeLostTeamAgent()
            }

            override fun action() {
                agent.spendTeamAgentFuel(3)
            }
        })
        // -------------------------------------------->
        createRule(object : Rule(800, "Fuel") {
            override fun constraint(): Boolean {
                return agent.isAtMothership && (agent.fuelInPercent <= 50)
            }

            override fun action() {
                agent.orderFuel(100)
            }
        })
    }
}
